<a href="https://colab.research.google.com/github/KSK123123/AIFFEL_quest_cr/blob/master/Python/DA_project6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 6-1. 프로젝트 (1) load_digits : 손글씨를 분류해 봅시다

In [12]:
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

import numpy as np
import pandas as pd


In [2]:
# load_digits메서드를 사용하여 데이터 준비

digits = load_digits()

In [8]:
# Feature Data 지정하기
X = digits.data
X

# Label Data 지정하기
Y = digits.target
Y

# Target Names 출력해 보기
Y_name = digits.target_names
Y_name

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [14]:
# 데이터 Describe 해 보기
digits_df = pd.DataFrame(data=digits.data, columns=digits.feature_names)
digits_df['label'] = digits.target
digits_df.describe()

pixel_0_0    pixel_0_1    pixel_0_2    pixel_0_3    pixel_0_4  \
count     1797.0  1797.000000  1797.000000  1797.000000  1797.000000   
mean         0.0     0.303840     5.204786    11.835838    11.848080   
std          0.0     0.907192     4.754826     4.248842     4.287388   
min          0.0     0.000000     0.000000     0.000000     0.000000   
25%          0.0     0.000000     1.000000    10.000000    10.000000   
50%          0.0     0.000000     4.000000    13.000000    13.000000   
75%          0.0     0.000000     9.000000    15.000000    15.000000   
max          0.0     8.000000    16.000000    16.000000    16.000000   

         pixel_0_5    pixel_0_6    pixel_0_7    pixel_1_0    pixel_1_1  ...  \
count  1797.000000  1797.000000  1797.000000  1797.000000  1797.000000  ...   
mean      5.781859     1.362270     0.129661     0.005565     1.993879  ...   
std       5.666418     3.325775     1.037383     0.094222     3.196160  ...   
min       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
25%       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
50%       4.000000     0.000000     0.000000     0.000000     0.000000  ...   
75%      11.000000     0.000000     0.000000     0.000000     3.000000  ...   
max      16.000000    16.000000    15.000000     2.000000    16.000000  ...   

         pixel_6_7    pixel_7_0    pixel_7_1    pixel_7_2    pixel_7_3  \
count  1797.000000  1797.000000  1797.000000  1797.000000  1797.000000   
mean      0.206455     0.000556     0.279354     5.557596    12.089037   
std       0.984401     0.023590     0.934302     5.103019     4.374694   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     1.000000    11.000000   
50%       0.000000     0.000000     0.000000     4.000000    13.000000   
75%       0.000000     0.000000     0.000000    10.000000    16.000000   
max      13.000000     1.000000     9.000000    16.000000    16.000000   

         pixel_7_4    pixel_7_5    pixel_7_6    pixel_7_7        label  
count  1797.000000  1797.000000  1797.000000  1797.000000  1797.000000  
mean     11.809126     6.764051     2.067891     0.364496     4.490818  
std       4.933947     5.900623     4.090548     1.860122     2.865304  
min       0.000000     0.000000     0.000000     0.000000     0.000000  
25%      10.000000     0.000000     0.000000     0.000000     2.000000  
50%      14.000000     6.000000     0.000000     0.000000     4.000000  
75%      16.000000    12.000000     2.000000     0.000000     7.000000  
max      16.000000    16.000000    16.000000    16.000000     9.000000  

[8 rows x 65 columns]

In [15]:
# train, test 데이터 분리

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=23)

In [28]:
# 5. 다양한 모델로 학습 및 6. 평
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.linear_model import SGDClassifier, LogisticRegression
from sklearn.metrics import classification_report, accuracy_score

In [29]:
# Decision Tree
decision_tree = DecisionTreeClassifier(random_state=42)
decision_tree.fit(X_train, Y_train)

Y_pred_dt = decision_tree.predict(X_test)
print("Decision Tree 성능 평가")
print(classification_report(Y_test, Y_pred_dt))
print(f"Accuracy: {accuracy_score(Y_test, Y_pred_dt)}\n")

Decision Tree 성능 평가
              precision    recall  f1-score   support

           0       0.88      0.97      0.92        30
           1       0.94      0.80      0.86        40
           2       0.94      0.87      0.90        38
           3       0.73      0.83      0.77        29
           4       0.92      0.85      0.88        40
           5       0.86      0.84      0.85        44
           6       0.90      0.90      0.90        31
           7       0.79      0.93      0.85        28
           8       0.78      0.82      0.79        38
           9       0.80      0.79      0.80        42

    accuracy                           0.85       360
   macro avg       0.85      0.86      0.85       360
weighted avg       0.86      0.85      0.85       360

Accuracy: 0.8527777777777777



In [30]:
# Random Forest
random_forest = RandomForestClassifier(random_state=42)
random_forest.fit(X_train, Y_train)

Y_pred_rf = random_forest.predict(X_test)
print("Random Forest 성능 평가")
print(classification_report(Y_test, Y_pred_rf))
print(f"Accuracy: {accuracy_score(Y_test, Y_pred_rf)}\n")

Random Forest 성능 평가
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        30
           1       0.97      0.97      0.97        40
           2       1.00      1.00      1.00        38
           3       0.97      1.00      0.98        29
           4       0.97      0.97      0.97        40
           5       0.93      0.98      0.96        44
           6       1.00      1.00      1.00        31
           7       0.97      1.00      0.98        28
           8       0.97      0.95      0.96        38
           9       1.00      0.93      0.96        42

    accuracy                           0.98       360
   macro avg       0.98      0.98      0.98       360
weighted avg       0.98      0.98      0.98       360

Accuracy: 0.9777777777777777



In [31]:
# Support Vector Machine (SVM)
svm = SVC(random_state=42)
svm.fit(X_train, Y_train)

Y_pred_svm = svm.predict(X_test)
print("SVM 성능 평가")
print(classification_report(Y_test, Y_pred_svm))
print(f"Accuracy: {accuracy_score(Y_test, Y_pred_svm)}\n")

SVM 성능 평가
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        30
           1       0.98      1.00      0.99        40
           2       1.00      1.00      1.00        38
           3       1.00      1.00      1.00        29
           4       1.00      0.97      0.99        40
           5       1.00      1.00      1.00        44
           6       1.00      1.00      1.00        31
           7       1.00      1.00      1.00        28
           8       0.95      0.97      0.96        38
           9       1.00      0.98      0.99        42

    accuracy                           0.99       360
   macro avg       0.99      0.99      0.99       360
weighted avg       0.99      0.99      0.99       360

Accuracy: 0.9916666666666667



In [32]:
# SGD Classifier
sgd = SGDClassifier(random_state=42)
sgd.fit(X_train, Y_train)

Y_pred_sgd = sgd.predict(X_test)
print("SGD Classifier 성능 평가")
print(classification_report(Y_test, Y_pred_sgd))
print(f"Accuracy: {accuracy_score(Y_test, Y_pred_sgd)}\n")

SGD Classifier 성능 평가
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        30
           1       0.88      0.90      0.89        40
           2       1.00      0.97      0.99        38
           3       0.72      1.00      0.84        29
           4       1.00      0.95      0.97        40
           5       0.98      0.95      0.97        44
           6       1.00      0.97      0.98        31
           7       0.87      0.96      0.92        28
           8       0.97      0.84      0.90        38
           9       1.00      0.88      0.94        42

    accuracy                           0.94       360
   macro avg       0.94      0.94      0.94       360
weighted avg       0.95      0.94      0.94       360

Accuracy: 0.9388888888888889



In [33]:
# Logistic Regression
logistic_regression = LogisticRegression(max_iter=10000, random_state=42)
logistic_regression.fit(X_train, Y_train)

Y_pred_lr = logistic_regression.predict(X_test)
print("Logistic Regression 성능 평가")
print(classification_report(Y_test, Y_pred_lr))
print(f"Accuracy: {accuracy_score(Y_test, Y_pred_lr)}\n")

Logistic Regression 성능 평가
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        30
           1       0.89      0.97      0.93        40
           2       1.00      1.00      1.00        38
           3       0.97      1.00      0.98        29
           4       1.00      0.95      0.97        40
           5       0.93      0.95      0.94        44
           6       0.97      0.97      0.97        31
           7       0.93      1.00      0.97        28
           8       0.97      0.87      0.92        38
           9       1.00      0.95      0.98        42

    accuracy                           0.96       360
   macro avg       0.97      0.97      0.97       360
weighted avg       0.97      0.96      0.96       360

Accuracy: 0.9638888888888889



#**각 모델의 성능 해석**
###Decision Tree 성능 평가

정확도: 85.28%\
정밀도: 0.73에서 0.94까지 다양하다.
숫자 3과 8에서 정밀도가 상대적으로 낮음.\
재현율: 0.79에서 0.97까지 다양하다.(1, 2, 3, 4, 5, 8)낮은 클래스가 많다.\
F1-Score: 0.77에서 0.92까지 다양하다.
3, 8에서 상대적으로 낮음.\
\
해석: 전반적으로 괜찮은 성능을 보이지만 숫자 3과 8에서 성능이 떨어짐.

---
###Random Forest 성능 평가

정확도: 97.78%\
정밀도: 거의 1.00으로 매우 높음.\
재현율: 거의 0.93 이상으로 매우 높음.\
F1-Score: 거의 0.96 이상으로 매우 높음.\
\
해석: 모든 숫자에서 매우 높은 성능을 보인다.


---
###SVM 성능 평가

정확도: 99.17%\
정밀도: 거의 1.00으로 매우 높음.\
재현율: 거의 1.00으로 매우 높음.\
F1-Score: 거의 1.00으로 매우 높음.\
해석: 손글씨 분류에서 거의 완벽에 가까운 성능을 보임.

---
###SGD Classifier 성능 평가

정확도: 93.89%
정밀도: 대부분의 숫자에서 0.84 이상. 그러나 숫자 3에서 낮은 편(0.72)\
재현율: 대부분 0.95 이상으로 높으나, 1과 8, 9에서 상대적으로 약간 낮긴하다.
F1-Score: 대부분의 숫자에서 0.84 이상으로 양호함.
해석: 전반적으로 양호한 성능을 보임.

---
###Logistic Regression 성능 평가

정확도: 96.39%\
정밀도: 대부분 0.89 이상.\
재현율: 대부분 0.95 이상.\
F1-Score: 0.92~1.00
해석: 매우 좋은 성능을 보인다.

---
## **결론**
SVM이 거의 100%에 가까운 성능을 보여주며 가장 적합한 모델임을 알 수 있음.

---
# 6-2. 프로젝트 (2) load_wine : 와인을 분류해 봅시다

In [36]:
# 1. 필요한 라이브러리 import
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

In [38]:
# 2. 데이터셋 준비
wine = load_wine()

In [58]:
# 3. 데이터 이해
# Feature Data (와인 특성 데이터)
X = wine.data

# Label Data (와인 레이블 데이터)
Y = wine.target

# Target Names (클래스 레이블)
target_names = wine.target_names
target_names

array(['class_0', 'class_1', 'class_2'], dtype='<U7')

In [59]:
# 데이터 통계 정보 출력
wine_df = pd.DataFrame(X, columns=wine.feature_names)
wine_df.describe()

alcohol  malic_acid         ash  alcalinity_of_ash   magnesium  \
count  178.000000  178.000000  178.000000         178.000000  178.000000   
mean    13.000618    2.336348    2.366517          19.494944   99.741573   
std      0.811827    1.117146    0.274344           3.339564   14.282484   
min     11.030000    0.740000    1.360000          10.600000   70.000000   
25%     12.362500    1.602500    2.210000          17.200000   88.000000   
50%     13.050000    1.865000    2.360000          19.500000   98.000000   
75%     13.677500    3.082500    2.557500          21.500000  107.000000   
max     14.830000    5.800000    3.230000          30.000000  162.000000   

       total_phenols  flavanoids  nonflavanoid_phenols  proanthocyanins  \
count     178.000000  178.000000            178.000000       178.000000   
mean        2.295112    2.029270              0.361854         1.590899   
std         0.625851    0.998859              0.124453         0.572359   
min         0.980000    0.340000              0.130000         0.410000   
25%         1.742500    1.205000              0.270000         1.250000   
50%         2.355000    2.135000              0.340000         1.555000   
75%         2.800000    2.875000              0.437500         1.950000   
max         3.880000    5.080000              0.660000         3.580000   

       color_intensity         hue  od280/od315_of_diluted_wines      proline  
count       178.000000  178.000000                    178.000000   178.000000  
mean          5.058090    0.957449                      2.611685   746.893258  
std           2.318286    0.228572                      0.709990   314.907474  
min           1.280000    0.480000                      1.270000   278.000000  
25%           3.220000    0.782500                      1.937500   500.500000  
50%           4.690000    0.965000                      2.780000   673.500000  
75%           6.200000    1.120000                      3.170000   985.000000  
max          13.000000    1.710000                      4.000000  1680.000000

In [42]:
# 4. train, test 데이터 분리
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.3, random_state=42)


In [43]:
# 1. Decision Tree
decision_tree = DecisionTreeClassifier(random_state=42)
decision_tree.fit(X_train, Y_train)

Y_pred_dt = decision_tree.predict(X_test)
print("Decision Tree 성능 평가")
print(classification_report(Y_test, Y_pred_dt))
print(f"Accuracy: {accuracy_score(Y_test, Y_pred_dt)}\n")

Decision Tree 성능 평가
              precision    recall  f1-score   support

           0       0.95      0.95      0.95        19
           1       0.95      1.00      0.98        21
           2       1.00      0.93      0.96        14

    accuracy                           0.96        54
   macro avg       0.97      0.96      0.96        54
weighted avg       0.96      0.96      0.96        54

Accuracy: 0.9629629629629629



In [44]:
# 2. Random Forest
random_forest = RandomForestClassifier(random_state=42)
random_forest.fit(X_train, Y_train)

Y_pred_rf = random_forest.predict(X_test)
print("Random Forest 성능 평가")
print(classification_report(Y_test, Y_pred_rf))
print(f"Accuracy: {accuracy_score(Y_test, Y_pred_rf)}\n")

Random Forest 성능 평가
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        19
           1       1.00      1.00      1.00        21
           2       1.00      1.00      1.00        14

    accuracy                           1.00        54
   macro avg       1.00      1.00      1.00        54
weighted avg       1.00      1.00      1.00        54

Accuracy: 1.0



In [45]:
# 3. Support Vector Machine (SVM)
svm = SVC(random_state=42)
svm.fit(X_train, Y_train)

Y_pred_svm = svm.predict(X_test)
print("SVM 성능 평가")
print(classification_report(Y_test, Y_pred_svm))
print(f"Accuracy: {accuracy_score(Y_test, Y_pred_svm)}\n")

SVM 성능 평가
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        19
           1       0.63      0.90      0.75        21
           2       0.60      0.21      0.32        14

    accuracy                           0.76        54
   macro avg       0.74      0.71      0.69        54
weighted avg       0.75      0.76      0.72        54

Accuracy: 0.7592592592592593



In [46]:
# 4. SGD Classifier
sgd = SGDClassifier(random_state=42)
sgd.fit(X_train, Y_train)

Y_pred_sgd = sgd.predict(X_test)
print("SGD Classifier 성능 평가")
print(classification_report(Y_test, Y_pred_sgd))
print(f"Accuracy: {accuracy_score(Y_test, Y_pred_sgd)}\n")

SGD Classifier 성능 평가
              precision    recall  f1-score   support

           0       0.50      1.00      0.67        19
           1       0.69      0.52      0.59        21
           2       0.00      0.00      0.00        14

    accuracy                           0.56        54
   macro avg       0.40      0.51      0.42        54
weighted avg       0.44      0.56      0.47        54

Accuracy: 0.5555555555555556



/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [47]:
# 5. Logistic Regression
logistic_regression = LogisticRegression(max_iter=10000, random_state=42)
logistic_regression.fit(X_train, Y_train)

Y_pred_lr = logistic_regression.predict(X_test)
print("Logistic Regression 성능 평가")
print(classification_report(Y_test, Y_pred_lr))
print(f"Accuracy: {accuracy_score(Y_test, Y_pred_lr)}\n")

Logistic Regression 성능 평가
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        19
           1       1.00      1.00      1.00        21
           2       1.00      1.00      1.00        14

    accuracy                           1.00        54
   macro avg       1.00      1.00      1.00        54
weighted avg       1.00      1.00      1.00        54

Accuracy: 1.0



#**각 모델의 성능 해석**
###Decision Tree 성능 평가

정확도: 96.3%\
정밀도: 0.95 이상\
재현율: 0.93 이상\
F1-Score: 0.95 이상\
해석: 전반적으로 높은 성능

---
###Random Forest 성능 평가

정확도: 100%\
정밀도: 모든 클래스 1.00\
재현율: 모든 클래스 1.00\
F1-Score: 모든 클래스 1.00\
해석: 모든 샘플을 정확하게 예측함.

---
###SVM 성능 평가

정확도: 75.9%\
정밀도: 클래스 0은 1.00, 클래스 1은 0.63, 클래스 2는 0.60\
재현율: 클래스 0은 1.00, 클래스 1은 0.90, 클래스 2는 0.21\
F1-Score: 클래스 0에서 1.00으로 좋으나, 클래스 2에서 낮은 0.32\
해석: 클래스 0에 대해 매우 좋은 성능을 보이지만, 클래스 1과 2에 대해선 저조한 성능을 보여 해당 모델은 데이터에 적합하지 않아 보임.

---
###SGD Classifier 성능 평가

정확도: 55.6%\
정밀도: 클래스 0은 0.50, 클래스 1은 0.69, 클래스 2는 0.00\
재현율: 클래스 0은 1.00, 클래스 1은 0.52, 클래스 2는 0.00\
F1-Score: 클래스 0에서 0.67로 양호하지만, 클래스 2는 0.00\
해석: 클래스 2를 전혀 예측하지 못하고 있다.해당 모델은 데이터에 적합하지 않아 보임.

---
###Logistic Regression 성능 평가

정확도: 100%\
정밀도: 모든 클래스에서 1.00\
재현율: 모든 클래스에서 1.00\
F1-Score: 모든 클래스에서 1.00\
해석: Random Forest와 마찬가지로 완벽한 성능을 보인다.

---
## 결론
Random Forest, Logistic Regression이 모든 클래스를 정확하게 예측하여 가장 적합한 모델이라는 것을 확인할 수 있다.
그러나 이는 과적합의 가능성도 고려해야 한다.

---
# 6-3. 프로젝트 (3) load_breast_cancer : 유방암 여부를 진단해 봅시다

In [48]:
from sklearn.datasets import load_breast_cancer

In [49]:
breast_cancer = load_breast_cancer()

In [61]:
X = breast_cancer.data
Y = breast_cancer.target

target_names = breast_cancer.target_names

breast_cancer_df = pd.DataFrame(X, columns=breast_cancer.feature_names)
breast_cancer_df.describe()

mean radius  mean texture  mean perimeter    mean area  \
count   569.000000    569.000000      569.000000   569.000000   
mean     14.127292     19.289649       91.969033   654.889104   
std       3.524049      4.301036       24.298981   351.914129   
min       6.981000      9.710000       43.790000   143.500000   
25%      11.700000     16.170000       75.170000   420.300000   
50%      13.370000     18.840000       86.240000   551.100000   
75%      15.780000     21.800000      104.100000   782.700000   
max      28.110000     39.280000      188.500000  2501.000000   

       mean smoothness  mean compactness  mean concavity  mean concave points  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       mean symmetry  mean fractal dimension  ...  worst radius  \
count     569.000000              569.000000  ...    569.000000   
mean        0.181162                0.062798  ...     16.269190   
std         0.027414                0.007060  ...      4.833242   
min         0.106000                0.049960  ...      7.930000   
25%         0.161900                0.057700  ...     13.010000   
50%         0.179200                0.061540  ...     14.970000   
75%         0.195700                0.066120  ...     18.790000   
max         0.304000                0.097440  ...     36.040000   

       worst texture  worst perimeter   worst area  worst smoothness  \
count     569.000000       569.000000   569.000000        569.000000   
mean       25.677223       107.261213   880.583128          0.132369   
std         6.146258        33.602542   569.356993          0.022832   
min        12.020000        50.410000   185.200000          0.071170   
25%        21.080000        84.110000   515.300000          0.116600   
50%        25.410000        97.660000   686.500000          0.131300   
75%        29.720000       125.400000  1084.000000          0.146000   
max        49.540000       251.200000  4254.000000          0.222600   

       worst compactness  worst concavity  worst concave points  \
count         569.000000       569.000000            569.000000   
mean            0.254265         0.272188              0.114606   
std             0.157336         0.208624              0.065732   
min             0.027290         0.000000              0.000000   
25%             0.147200         0.114500              0.064930   
50%             0.211900         0.226700              0.099930   
75%             0.339100         0.382900              0.161400   
max             1.058000         1.252000              0.291000   

       worst symmetry  worst fractal dimension  
count      569.000000               569.000000  
mean         0.290076                 0.083946  
std          0.061867                 0.018061  
min          0.156500                 0.055040  
25%          0.250400                 0.071460  
50%          0.282200                 0.080040  
75%          0.317900                 0.092080  
max          0.663800                 0.207500  

[8 rows x 30 columns]

In [51]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.3, random_state=42)

In [52]:
# 1. Decision Tree
decision_tree = DecisionTreeClassifier(random_state=42)
decision_tree.fit(X_train, Y_train)

Y_pred_dt = decision_tree.predict(X_test)
print("Decision Tree 성능 평가")
print(classification_report(Y_test, Y_pred_dt))
print(f"Accuracy: {accuracy_score(Y_test, Y_pred_dt)}\n")

Decision Tree 성능 평가
              precision    recall  f1-score   support

           0       0.90      0.95      0.92        63
           1       0.97      0.94      0.95       108

    accuracy                           0.94       171
   macro avg       0.93      0.94      0.94       171
weighted avg       0.94      0.94      0.94       171

Accuracy: 0.9415204678362573



In [53]:
# 2. Random Forest
random_forest = RandomForestClassifier(random_state=42)
random_forest.fit(X_train, Y_train)

Y_pred_rf = random_forest.predict(X_test)
print("Random Forest 성능 평가")
print(classification_report(Y_test, Y_pred_rf))
print(f"Accuracy: {accuracy_score(Y_test, Y_pred_rf)}\n")

Random Forest 성능 평가
              precision    recall  f1-score   support

           0       0.98      0.94      0.96        63
           1       0.96      0.99      0.98       108

    accuracy                           0.97       171
   macro avg       0.97      0.96      0.97       171
weighted avg       0.97      0.97      0.97       171

Accuracy: 0.9707602339181286



In [54]:
# 3. Support Vector Machine (SVM)
svm = SVC(random_state=42)
svm.fit(X_train, Y_train)

Y_pred_svm = svm.predict(X_test)
print("SVM 성능 평가")
print(classification_report(Y_test, Y_pred_svm))
print(f"Accuracy: {accuracy_score(Y_test, Y_pred_svm)}\n")

SVM 성능 평가
              precision    recall  f1-score   support

           0       1.00      0.83      0.90        63
           1       0.91      1.00      0.95       108

    accuracy                           0.94       171
   macro avg       0.95      0.91      0.93       171
weighted avg       0.94      0.94      0.93       171

Accuracy: 0.935672514619883



In [55]:
# 4. SGD Classifier
sgd = SGDClassifier(random_state=42)
sgd.fit(X_train, Y_train)

Y_pred_sgd = sgd.predict(X_test)
print("SGD Classifier 성능 평가")
print(classification_report(Y_test, Y_pred_sgd))
print(f"Accuracy: {accuracy_score(Y_test, Y_pred_sgd)}\n")

SGD Classifier 성능 평가
              precision    recall  f1-score   support

           0       1.00      0.86      0.92        63
           1       0.92      1.00      0.96       108

    accuracy                           0.95       171
   macro avg       0.96      0.93      0.94       171
weighted avg       0.95      0.95      0.95       171

Accuracy: 0.9473684210526315



In [56]:
# 5. Logistic Regression
logistic_regression = LogisticRegression(max_iter=10000, random_state=42)
logistic_regression.fit(X_train, Y_train)

Y_pred_lr = logistic_regression.predict(X_test)
print("Logistic Regression 성능 평가")
print(classification_report(Y_test, Y_pred_lr))
print(f"Accuracy: {accuracy_score(Y_test, Y_pred_lr)}\n")

Logistic Regression 성능 평가
              precision    recall  f1-score   support

           0       0.97      0.97      0.97        63
           1       0.98      0.98      0.98       108

    accuracy                           0.98       171
   macro avg       0.97      0.97      0.97       171
weighted avg       0.98      0.98      0.98       171

Accuracy: 0.9766081871345029



# 각 모델의 성능 해석
###Decision Tree 성능 평가

정확도: 94.15%\
정밀도: 0.90 (악성 종양), 0.97 (양성 종양)\
재현율: 0.95 (악성 종양), 0.94 (양성 종양)\
F1-Score: 전체적으로 0.92 이상으로 높은 수준\
해석: 전체적으로 양호한 성능을 보이지만, 악성 종양(클래스 0)에 대해 상대적으로 재현율이 약간 낮다.

---
###Random Forest 성능 평가

정확도: 97.08%\
정밀도: 0.98 (악성 종양), 0.96 (양성 종양)\
재현율: 0.94 (악성 종양), 0.99 (양성 종양)\
F1-Score: 0.96 이상\
해석: 전체적으로 양호한 성능을 보임

---
###SVM 성능 평가

정확도: 93.56%\
정밀도: 1.00 (악성 종양), 0.91 (양성 종양)\
재현율: 0.83 (악성 종양), 1.00 (양성 종양)\
F1-Score: 0.90 이상\
해석: 양성 종양을 잘 분류하지만, 악성 종양(클래스 0)에 대한 재현율이 낮아 악성 종양을 놓칠 위험이 있다.

---
###SGD Classifier 성능 평가

정확도: 94.74%\
정밀도: 1.00 (악성 종양), 0.92 (양성 종양)\
재현율: 0.86 (악성 종양), 1.00 (양성 종양)\
F1-Score: 0.92 이상\
해석: SVM과 비슷하게 악성 종양에 대한 재현율이 낮다.

---
###Logistic Regression 성능 평가

정확도: 97.66%\
정밀도: 0.97 (악성 종양), 0.98 (양성 종양)\
재현율: 0.97 (악성 종양), 0.98 (양성 종양)\
F1-Score: 0.97 이상으로 매우 좋음\
해석: 전체적으로 매우 좋은 성능을 보여준다.

---
## 결론
유방암 여부 진단에서 가장 먼저 고려해야할 것은 **재현율(Recall)**이다.\
이는 종양 분류에서는 악성 종양을 놓치는 위험을 최소화하고자 하기 때문이다.

따라서 재현율도 높고 전체적으로 성능이 좋은 Logistic Regression이 가장 적합하다고 볼 수 있다.